### PR TITLE
V4 bridge upgrade

### PR DESCRIPTION
**Pull request by bravepooh**
Features in this PR:
1. Enable owner (multisig) to mark transaction as not processed and not voted.
This will allow federators to vote and execute tx that were stucked on the bridge.
2. Enable mint of ERC777 SideToken to a smart contract that doesn't have ERC777 interface.
This will be done by erc777Converter that will receive the funds and transfer.

### CODE REVIEW

File: sovryn-token-bridge/bridge/contracts/Bridge.sol [hunk 1]
```diff
@@ -168,18 +169,25 @@ contract Bridge is Initializable, IBridge, IERC777Recipient, UpgradablePausable,
...
-            (bool success, bytes memory errorData) = receiver.call(
-                abi.encodeWithSignature("onTokensMinted(uint256,address,bytes)", formattedAmount, sideToken, userData)
-            );
-            if (!success) {
-                emit ErrorTokenReceiver(errorData);
-            }
-        }
+    // Enable mint to SC without ERC777 interface by using erc777Converter
+        address receiverU = receiver;
+        bytes memory userDataU = userData;
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol]: Removed

File: sovryn-token-bridge/bridge/contracts/Bridge.sol [hunk 2]
```diff
@@ -168,18 +169,25 @@ contract Bridge is Initializable, IBridge, IERC777Recipient, UpgradablePausable,
...
+            receiverU = erc777ConverterAddr;
+        }
+    
+        sideToken.mint(receiverU, formattedAmount, userDataU, "");    
+
+        if (receiverU.isContract()) {
+           (bool success, bytes memory errorData) = receiverU.call(
+               abi.encodeWithSignature("onTokensMinted(uint256,address,bytes)", formattedAmount, sideToken, userDataU)
+           );
+           emit ErrorTokenReceiver(errorData);
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol]: Removed

File: sovryn-token-bridge/bridge/contracts/Bridge_v3.sol [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+    //         crossTokens(WETHAddr, msg.sender, msg.value, _userData);
+    //     }
+    // }
+    // function setAggregatorAddr(address _aggregatorAddr) external onlyOwner {
+    //     aggregatorAddr = _aggregatorAddr;
+    // }
+
+    function setWETHAddress(address _WETHAddr) external onlyOwner {
+        require(_WETHAddr != address(0) && !ethFirstTransfer , "No set WETHAddr AF 1st transfer");
+        //require(!ethFirstTransfer, "cannot change WETHAddr after first transfer");
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge_v3.sol:436]: Removed

File: sovryn-token-bridge/bridge/contracts/Bridge.sol [hunk 4]
```diff
@@ -468,6 +476,16 @@ contract Bridge is Initializable, IBridge, IERC777Recipient, UpgradablePausable,
         return nativeTokenSymbol;
     }
 
+    function setRevokeTransaction(bytes32 _revokeTransactionID) external onlyOwner {
+        require(_revokeTransactionID != NULL_HASH, "_revokeTransactionID cannot be NULL");
+        processed[_revokeTransactionID] = false;
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol:483]: Done

File: sovryn-token-bridge/bridge/contracts/Bridge_v3.sol [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+        require(!initialPrefixSetup, "Bridge: initialPrefixSetup Done");
+        isSuffix = _isSuffix;
+        symbolPrefix = _prefix;
+        //emit PrefixUpdated(isSuffix, _prefix);
+    }
+
+    function withdrawAllEthFees(address payable _to) public payable onlyOwner {
+        require(address(this).balance >= ethFeeCollected);
+        ethFeeCollected = 0;
+        _to.transfer(ethFeeCollected);
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge_v3.sol:460]: Changed implementation:
        require(address(this).balance >= ethFeeCollected);
        uint256 sendEthFeeCollected = ethFeeCollected;
        ethFeeCollected = 0;
        _to.transfer(sendEthFeeCollected);

File: sovryn-token-bridge/bridge/contracts/Bridge_v3.sol [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    function setNativeTokenSymbol(string calldata _nativeTokenSymbol) external onlyOwner {
+        nativeTokenSymbol = _nativeTokenSymbol;
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Bridge_v3.sol:464]: DONE

File: sovryn-token-bridge/bridge/contracts/Federation.sol [hunk 7]
```diff
@@ -205,4 +207,14 @@ contract Federation is Ownable {
         emit RequirementChange(_required);
     }
 
+    function setRevokeTransactionAndVote(bytes32 _revokeTransactionID) external onlyOwner {
+        require(_revokeTransactionID != NULL_HASH, "_revokeTransactionID cannot be NULL");
+        
+        processed[_revokeTransactionID] = false;
```
[the remark this answers is not in the document]
Comment (bravepooh) [reply sovryn-token-bridge/bridge/contracts/Federation.sol:213]: Done

File: sovryn-token-bridge/bridge/contracts/Bridge.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajaroy43) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol:483]: Check here if specific tx id id processed or not 
      `  require(processed[_revokeTransactionID],"Bridge: Tx id not processed  "); `

File: sovryn-token-bridge/bridge/contracts/Federation.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (rajaroy43) [reply sovryn-token-bridge/bridge/contracts/Federation.sol:213]: probable ,we need same - 
Check here if specific tx id processed or not
require(processed[_revokeTransactionID],"Federation": Tx id not processed ");

File: sovryn-token-bridge/bridge/contracts/Bridge.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavidHaller43) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol:476]: i suggest to check that specific txid is processed or not   `require(processed[_revokeTransactionID], "Bridge: cannot revoke unprocessed TX");   `

[the remark this answers is not in the document]
Comment (rajaroy43) [reply sovryn-token-bridge/bridge/contracts/Bridge.sol:476]: i suggest to check that specific txid is processed or not   `require(processed[_revokeTransactionID], "Bridge: cannot revoke unprocessed TX");   `

File: sovryn-token-bridge/bridge/contracts/Federation.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (rajaroy43) [reply sovryn-token-bridge/bridge/contracts/Federation.sol:214]: probably same we need here .
`require(processed[_revokeTransactionID], "Federation: cannot revoke unprocessed TX");   `

**Comment by jjmr007** [comment]
One of the affected users wants to know the status of this PR. What is the updated ETA for this changes to be merged? Thanks

**Pull request merged**